### PR TITLE
feat: integrate SIMD into OAHSet

### PR DESCRIPTION
**Pull request by BorysTheDev**
Added a wrapper for the vector compiler extension that was integrated into the OAHSet::Add() method.
Add a regression test ensuring re-adding keys never creates duplicates across rehashes and vector overflows.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -109,6 +109,29 @@ endif()
 
 include(third_party)
 include(internal)
+# helio's internal.cmake sets CMAKE_CXX_STANDARD=17 for its own legacy targets;
+# dragonfly itself uses C++20 (operator<=>, std::contiguous_iterator, concepts),
+# so restore the C++20 standard before configuring our own subdirectories.
+set(CMAKE_CXX_STANDARD 20)
```
Comment (romange) [inline CMakeLists.txt]: And I would like to avoid this as well. Or at least to take it slow at my own pace.

File: CMakeLists.txt [hunk 2]
```diff
@@ -109,6 +109,29 @@ endif()
...
+
+# Fetch the EVE (Expressive Vector Engine) SIMD library used by OAHSet.
+# Mirrors the FetchContent pattern in helio/cmake/third_party.cmake.
+include(FetchContent)
+FetchContent_Declare(
+  eve
+  URL https://github.com/jfalcou/eve/archive/refs/tags/v2023.02.15.tar.gz
```
Comment (romange) [inline CMakeLists.txt]: I looked at it and I would like to avoid it.

1. It's a research project as they say themselves.
2. We already have `src/core/sse_port.h` which includes low-level instrinics for both x64 and arm64.
   I know that this is not the same but is it really hard to implement a few high level routines that you use via instrinics directly?

File: src/core/oah_set.h [hunk 3]
```diff
@@ -8,7 +8,9 @@
 #include <absl/random/random.h>
 #include <absl/types/span.h>
 
+#include <bit>
 #include <concepts>
+#include <eve/eve.hpp>
```
Comment (romange) [inline src/core/oah_set.h]: lets take more conservative approach and implement whatever semantics you need via self-contained code with low-level intrinsics.

File: CMakeLists.txt (hunk 2 above)
Comment (BorysTheDev) [reply CMakeLists.txt]: it is not hard, I even can do quite similar optimizations without intrinsic using compiler optimization logic and some compiler builtin functions. It will take some time and maybe be not so efficient, but quite OK.

Comment (romange) [reply CMakeLists.txt]: so the whole range of compiler instrinsics + sse/port include is to your disposal.

**Comment by BorysTheDev** [comment]
augment review

File: src/core/simd_op.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  SimdOp operator==(T value) const {  // NOLINT
+    return Vec(v_ == (Vec{} + value));
+  }
+
+  // The architecture branches exist because no portable formulation lowers to
+  // a single movemask instruction — every plain vector-extension version we
+  // tried measured ~5% slower on OAHSet's hot path.
+  BitsType ToBits() const {
```
Comment (romange) [inline src/core/simd_op.h]: it's not clear to me what it does.
if it only extracts LSB bits from the words, it's not a good name. Moreover it should have explanation here, and not in the class description.
The comments about the implementation should be in the function body, not before its declaration.

you mention LSB above but https://doc.rust-lang.org/stable/core/arch/x86_64/fn._mm256_movemask_pd.html

says it takes the MSB.

File: src/core/simd_op.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  using Vec __attribute__((vector_size(sizeof(T) * N))) = T;
+
+ public:
+  using BitsType = std::uint32_t;
+  static constexpr std::size_t kLanes = N;
+
+  SimdOp() = default;
+
+  // Broadcasting via `Vec{} + value` lowers to vpbroadcast on AVX2 / dup on
```
Comment (romange) [inline src/core/simd_op.h]: here as well. you talk about implementation but please explain what Broadcast does.

**Review by romange — COMMENTED** [review]
lgtm with minor comments

File: src/core/simd_op.h (hunk 5 above)
Comment (BorysTheDev) [reply src/core/simd_op.h]: Broadcast means to set one value to several elements in an array; it is standard SIMD terminology

File: src/core/simd_op.h (hunk 4 above)
Comment (BorysTheDev) [reply src/core/simd_op.h]: LSB is for a bitmask, because we store one bit for every element from 0 to 31. I will update comments

File: src/core/simd_op.h (hunk 5 above)
Comment (BorysTheDev) [reply src/core/simd_op.h]: renamed to Fill

File: src/core/simd_op.h (hunk 4 above)
Comment (BorysTheDev) [reply src/core/simd_op.h]: renamed and updated comments

File: src/core/simd_op.h [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  // NEON; a per-lane scalar loop pessimizes to vpinsrq + vperm2i128.
+  static SimdOp Fill(T value) {
+    return Vec{} + value;
+  }
+
+  static SimdOp Load(const T* ptr) {
+    Vec v;
+    std::memcpy(&v, ptr, sizeof(Vec));
+    return v;
+  }
```
Comment (dranikpg) [inline src/core/simd_op.h:47]: nit: you can use std::span or comment that it must be exactly the size of vec

File: src/core/simd_op.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    return Vec(v_ == o.v_);
+  }
+
+  SimdOp operator==(T value) const {  // NOLINT
+    return Vec(v_ == (Vec{} + value));
+  }
+
+  // Packs the most-significant bit of every lane into a uint32_t bitmask
+  // (LSB = lane 0). For the output of `operator==` (lanes are all-ones or
+  // all-zeros) this is equivalent to "bit i set iff lane i is non-zero".
```
Comment (dranikpg) [inline src/core/simd_op.h:75]: the "most-significant" part is a bit misleading, because in "bit i set iff lane i is non-zero" there is no most significant, but I understand that it's the same

File: src/core/oah_set.h [hunk 8]
```diff
@@ -136,33 +142,108 @@ class OAHSet {  // Open Addressing Hash Set
...
+      if (e.IsVector())
+        continue;
+      if (e.Key() != str) {  // after rehash, the pointer can miss hash so we need to set it for
+                             // better performance
+        if (e.GetHash() != ext_hash) {
+          e.SetExtHash(CalcExtHash(Hash(e.Key()), capacity_log_));
+        }
+        e.ExpireIfNeeded(time_now_, &size_, &obj_alloc_used_);
+        continue;
+      }
```
Comment (dranikpg) [inline src/core/oah_set.h:227]: looks really clean when simd parts are hidden 👍🏻

**Review by dranikpg — DISMISSED** [review]

File: src/core/oah_set.h (hunk 8 above)
Comment (BorysTheDev) [reply src/core/oah_set.h:227]: yes the algorithm is a little tricky because is built on bit magic, but it is a standard approach for SIMD so I think it should be clear

File: src/core/oah_set.h [hunk 9]
```diff
@@ -248,24 +329,45 @@ class OAHSet {  // Open Addressing Hash Set
...
-        entries_[bucket_id] = std::move(e);
-        return;
+  // Walk the vector at ext_bid (vectors live only at the extension point)
+  // looking for str. Returns nullptr if not found. Returns a pointer if found:
+  // caller checks .Empty() to distinguish "live match" (return false from Add)
+  // from "match-but-expired" (slot is now vacant; reuse it for the new entry).
+  //
+  // Vectors have power-of-2 capacity with minimum 2, so we sweep in 2-lane
+  // SIMD strides.
+  OAHEntry* ProbeExtensionVector(uint32_t ext_bid, std::string_view str, uint64_t ext_hash) {
```
Comment (romange) [inline src/core/oah_set.h:339]: I do not think that having those huge functions in header file helps with performance.

**Review by romange — APPROVED** [review]

File: src/core/oah_set.h (hunk 9 above)
Comment (BorysTheDev) [reply src/core/oah_set.h:339]: I will refactor it step by step
